src/gotcha_dl.c

```diff
@@ -54,14 +184,9 @@ static void* dlsym_wrapper(void* handle, const char* symbol_name){
   struct internal_binding_t *binding;
   int result;
   debug_printf(1, "User called dlsym(%p, %s)\n", handle, symbol_name);
-
   if(handle == RTLD_NEXT){
-    return _dl_sym(RTLD_NEXT, symbol_name ,__builtin_return_address(0));
+    return gotchas_dlsym_rtld_next_lookup(RTLD_NEXT, symbol_name ,__builtin_return_address(0));
   }
-  if(handle == RTLD_DEFAULT) {
-    return _dl_sym(RTLD_DEFAULT, symbol_name,__builtin_return_address(0));
-  }
-  
   result = lookup_hashtable(&function_hash_table, (hash_key_t) symbol_name, (hash_data_t *) &binding);
```

Comment:
Independent from your commit here (since this is in the original code)--I wonder if we should hoist the lookup_hashtable() call and test above the 'if (handle == RTLD_NEXT)' test. As we have this coded now, if the app calls dlsym(RTLD_NEXT, ...) or dlsym(RTLD_DEFAULT, ...) then it can go around gotcha's wrappers. 

I think we should be checking whether we wrap symbol_name, return the wrapper if so, then if not we can fall through into original dlsym behavior.  Not sure why it was originally coded like this.

Reply:
I agree. The logic of first checking wrapper seem more logical. Fixed the logic.